pymagicc/magicc_time.py

```diff
@@ -0,0 +1,201 @@
+            # middle of month
+            month = int(np.ceil(month_decimal))
+            _, month_days = monthrange(year, month)
+            day_decimal = month_days * 0.5
+            day = int(day_decimal)
+            hour = int(day_decimal % 1 * 24)
+        elif np.round(month_fraction, 1) == 0:
+            # start of month
+            month = int(month_decimal) + 1
+            day = 1
+            hour = 1
+        else:
+            error_msg = "Your timestamps don't appear to be middle or start of month"
+            raise ValueError(error_msg)
+
+        res = dt.datetime(year, month, day, hour)
+        return res
+
+    def convert_to_decimal_year(self, idtime):
+        """
+        Convert a datetime to MAGICC's expected decimal year representation
+
+        Parameters
+        ----------
+        idtime : :obj:`datetime.datetime`
+            :obj:`datetime.datetime` instance to convert to MAGICC's internal decimal
+            year conventions
+
+        Returns
+        -------
+        float
+            MAGICC's internal decimal year representation of ``idtime``
+
+        Raises
+        ------
+        ValueError
+            If we are not confident about how to convert the input times so that they
+            follow MAGICC's internal time conventions (i.e. we are not sure if
+            ``idtime`` is start or middle of the month).
+        """
+        year = idtime.year
+        month = idtime.month
+        year_fraction = (idtime - dt.datetime(year, 1, 1)).total_seconds() / (
+            dt.datetime(year + 1, 1, 1) - dt.datetime(year, 1, 1)
+        ).total_seconds()
+
+        midmonth_decimal_bit = ((month - 1) * 2 + 1) / 24
+        startmonth_decimal_bit = (month - 1) / 12
+        if self._yr_frac_close_to(year_fraction, self.midmonths_magicc):
+            decimal_bit = midmonth_decimal_bit
+        elif self._yr_frac_close_to(year_fraction, self.midmonths):
+            decimal_bit = midmonth_decimal_bit
+        elif self._yr_frac_close_to(
+            year_fraction, self.startmonths_magicc, must_be_greater=True
+        ):
+            decimal_bit = startmonth_decimal_bit
+        elif self._yr_frac_close_to(
+            year_fraction, self.startmonths, must_be_greater=True
+        ):
+            decimal_bit = startmonth_decimal_bit
+        else:
+            error_msg = "Your timestamps don't appear to be middle or start of month"
+            raise ValueError(error_msg)
+
+        return np.round(year + decimal_bit, 3)  # match MAGICC precision
+
+    def _yr_frac_close_to(self, yfrac, other, must_be_greater=False):
+        match_idx = np.where(
+            np.abs(yfrac - other) < self._convert_to_decimal_required_precision
+        )[0]
+        if match_idx.size == 0:
+            return False
+        if must_be_greater and yfrac < other[match_idx]:
+            return False
+        return True
+
+
+MAGICC_TIME_CONVERTER = _MAGICCTimeConverter()
```

Comment:
Why are you using a class here? purely for caching? Check out `functools.lru_cache`

I wouldn't capitalise MAGICC_TIME_CONVERTER as I tend to reserve that for constants where this is a singleton pattern. Pure stylistic choice and not sure what PEP8 says...

Reply:
Another solution would be to add the following lines so you can use `from pymagicc.magicc_time import convert_to_decimal_year`

```
convert_to_decimal_year = MAGICC_TIME_CONVERTER.convert_to_decimal_year
convert_to_datetime = MAGICC_TIME_CONVERTER.convert_to_datetime
```

Reply:
oops ye I should have used `functools.lru_cache` will change